### ISSUE TITLE
ControlPath in ssh_config fails v0.0.3.0

### ISSUE THREAD

**Issue by dougransom** [PowerShell/Win32-OpenSSH#405]
ssh config

Host blue
	HostName underdoug.ca
	User underdou
	ControlPath c:/users/doug/.ssh/controlmasters_doug_blue
	ControlMaster auto
    ControlPersist 10m

c:\Program Files\OpenSSH>ssh blue
muxclient socket(): Unknown error

**Comment by manojampalam** [comment]
ControlMaster is not supported on Windows yet. See https://github.com/PowerShell/Win32-OpenSSH/wiki/Project-Scope

**Comment by opoplawski** [comment]
Would be nice.

**Comment by dobsa** [comment]
It really would be nice. Please, bring it in!
I am using SSH that comes with latest Widows 10:

    OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5

and it still is not there :-(

**Comment by musm** [comment]
Also noted in https://github.com/PowerShell/openssh-portable/issues/463

**Comment by musm** [comment]
Setting in `~/.ssh/config`
```
ControlMaster auto
ControlPath ~/.ssh/master-%r@%h:%p
```

Leads to the errors when attempting to connect to any server:


```
getsockname failed: Bad file descriptor
channel_send_open: channel 0: unexpected internal error

```

**Comment by mcgarrah** [comment]
Still getting the error when trying the sockets for multiplexing.

```
C:\Users\mcgarrah>ssh -V
OpenSSH_for_Windows_7.6p1, LibreSSL 2.6.4
```

```
C:\Users\mcgarrah>ssh -S ~/.ssh/control/test  ec2-user@10.196.80.10
getsockname failed: Not a socket
```

```
C:\Users\mcgarrah>ssh -vvv -S ~/.ssh/control/test  ec2-user@10.196.80.10
OpenSSH_for_Windows_7.6p1, LibreSSL 2.6.4
debug1: Reading configuration data C:\\Users\\u107250/.ssh/config
debug3: Failed to open file:C:\\ProgramData\\ssh/ssh_config error:2
debug3: w32_getpeername ERROR: not sock :3
debug3: w32_getpeername ERROR: not sock :3
debug1: getpeername failed: Not a socket
debug3: w32_getsockname ERROR: not sock :3
debug3: w32_getsockname ERROR: not sock :3
getsockname failed: Not a socket
debug1: channel 0: new [client-session]
debug3: ssh_session2_open: channel_new: 0
debug2: channel 0: send open
debug3: ssh_packet_send_mux: type 90
debug1: Entering interactive session.
debug1: pledge: network
debug1: console supports the ansi parsing
debug2: client_check_window_change: changed
```

**Comment by mew1033** [comment]
Adding this would be amazing. It's the last thing (for me) that windows openssh doesn't have that things like Putty/Kitty do.

**Comment by Smithienious** [comment]
Wow that's more than 5 years already.

**Comment by juvieorg** [comment]
Hi man, I think that problem is in permission of mutiplex folder. The service account can't create there the socket.
If you add permission for multiplex folder problem solved (add SERVICE_ACCOUNT full control for 'control' folder in your case)
I tested on newer version OpenSSH in windoze - but issue started with same symptoms.
I hope the my note help you.

**Comment by solaris-ventus** [comment]
It's a pity, a pity. Required functionality

**Comment by mew1033** [comment]
Here's another ping for ControlMaster functionality. Any updated word on when/if this might be coming? Thanks!

**Comment by FlorianDeconinck** [comment]
Unreliable information, please disregard

~~Hi everyone.~~

~~Here's a workaround dug from the depth of [openssh bug list](https://anongit.mindrot.org/openssh.git/commit/?id=8d0578478586e283e751ca51e7b0690631da139a)~~

~~Tested on windows 10 2h22 / GitBash~~
~~* do your `ssh -MNf` with `ControlMaster` and `ControlPath` in the `config` unchanged~~
~~* then for each mux connection run `ssh -O proxy`~~

~~Does not fix original issue, but gives you mux terminal.~~

**Comment by woodda** [comment]
> add SERVICE_ACCOUNT full control for 'control' folder in your case

If this indeed is a workaround solution, it would be helpful if you gave step by step instructions for doing the above.

**Comment by youk** [comment]
@FlorianDeconinck Not clear how it is supposed to work and what you were testing. `-O` option requires setting ControlPath (`-S` option). Setting ControlPath makes OpenSSH fail, as it is not supported on Windows (this is what this issue is about).

**Comment by FlorianDeconinck** [comment]
> @FlorianDeconinck Not clear how it is supposed to work and what you were testing. `-O` option requires setting ControlPath (`-S` option). Setting ControlPath makes OpenSSH fail, as it is not supported on Windows (this is what this issue is about).

Hey,
I have actually moved on from this setup due too many problems stacking up, so indeed I might be talking out of nothing. I've edited the comment so it's disregarded.

Good luck on the bug hunt.

**Comment by youk** [comment]
@FlorianDeconinck Thanks. I don't think there is a bug. It's just not implemented. The only bug is that the corresponding command line options are available in the Windows port.

**Comment by timon-michel-scopevisio** [comment]
Can we get the "feature request" label specified on https://github.com/PowerShell/Win32-OpenSSH/wiki/Project-Scope here? I am afraid this issue might be forgotten because no one labeled it yet

**Comment by oladarula** [comment]
yeah, it is a pitty that it is not working. Seems the only feature holding me back for production use.

**Comment by kapitanluffy** [comment]
Assuming this will never be supported on Windows, any suggestions on an alternative instead?

**Comment by phiberoptick** [comment]
+1
Required functionality.  Especially the way Windows likes to reset network interfaces (VPN, HyperV etc)....

**Comment by draeath** [comment]
I find this ssh client useless without this functionality. Please prioritize.

**Comment by youk** [comment]
There are mommy and daddy for requests like this.

**Comment by rihards-simanovics** [comment]
> Assuming this will never be supported on Windows, any suggestions on an alternative instead?

Linux :)

**Comment by rihards-simanovics** [comment]
Jokes aside, I'm adding Windows support for the [git-sentinel](https://gitlab.com/griffin-web-studio/garage/git-sentinel) and that relies a lot on ControlMaster to ensure that a single ssh connection is re-used per host, in case one has many git repos on their PC. If I'm wrong in thinking there is no support, please let me know; otherwise I'll simply have to make ControlMaster unsupported for the app on Windows.

**Comment by duncm** [comment]
Had to reauthenticate one too many times so I threw some tokens at it: PowerShell/openssh-portable#863

I can see a few thumbs up here but I would recommend appropriate caution for anyone wanting to test it. The implementation necessarily differs from POSIX in multiple ways, so it will likely break in ways you don't expect.
